support/oidc-discovery-provider/healthchecks_handler.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"net/http"
+	"time"
+)
+
+const (
+	ThresholdMultiplicator = 5
+	ThresholdMinTime       = time.Minute * 3
+)
+
+type HealthChecksHandler struct {
+	source       JWKSSource
+	healthChecks HealthChecksConfig
+	jwkThreshold time.Duration
+	initTime     time.Time
+
+	http.Handler
+}
+
+func NewHealthChecksHandler(source JWKSSource, config *Config) *HealthChecksHandler {
+	h := &HealthChecksHandler{
+		source:       source,
+		healthChecks: *config.HealthChecks,
+		jwkThreshold: jwkThreshold(config),
+		initTime:     time.Now(),
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle(h.healthChecks.ReadyPath, http.HandlerFunc(h.readyCheck))
+	mux.Handle(h.healthChecks.LivePath, http.HandlerFunc(h.liveCheck))
+
+	h.Handler = mux
+	return h
+}
+
+// jwkThreshold determines the duration from the last successfull poll before the server is considered unhealthy
+func jwkThreshold(config *Config) time.Duration {
+	var duration time.Duration
+	if config.ServerAPI != nil {
+		duration = config.ServerAPI.PollInterval
+	} else {
+		duration = config.WorkloadAPI.PollInterval
+	}
+	if duration*ThresholdMultiplicator < ThresholdMinTime {
+		duration = ThresholdMinTime
+	}
+	return duration
+}
+
+// readyCheck is a health check that returns 200 if the server can successfully fetch a jwt keyset
+func (h *HealthChecksHandler) readyCheck(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	statusCode := http.StatusOK
+	_, modTime, isValid := h.source.FetchKeySet()
```

Comment:
Actually, `modTime` is _not_ the last time the source successfully polled for the JWKS, but rather the last time the JWKS changed. Therefore, it can't reliably be used to determine health since it is normal for the JWKS to not change frequently (it depends on how often servers are rotating their JWT signing keys).

Maybe the `Source` interface needs to grow a new method:
```
// LastPoll returns the time of the last successful poll of the JWKS from the source, or a zero value if
// there hasn't been a successful poll yet.
LastPoll() time.Time
```

Then each source can track and update the last time it successfully polled the source, whether or not there was an update. In terms of when they update, it could probably happen in each source's setJWKS function when either it determines there was no change, or when it updates the JWKS.

Reply:
Thanks for the hint. I missed that one :-)
Please review it again.